lib/Transformation.ts

```diff
@@ -156,25 +163,68 @@ function transformOperator(expr: Alg.OperatorExpression)
       throw new Err.UnknownOperator(expr.operator);
     }
     const op = expr.operator as C.RegularOperator;
-    const args = expr.args.map((a) => transformAlgebra(a));
+    const args = expr.args.map((a) => transformAlgebra(a, config));
     const func = regularFunctions.get(op);
     if (!hasCorrectArity(args, func.arity)) { throw new Err.InvalidArity(args, op); }
     return new E.Operator(args, func.apply);
   }
 }
 
+function wrapSyncFunction(f: SyncExtensionFunction, name: string): SimpleApplication {
+  return (args => {
+    let res;
+    try {
+      res = f(args.map(arg => arg.toRDF()));
+    }catch (e) {
+      throw new ExtensionFunctionError(name, e);
+    }
+    return transformRDFTermUnsafe(res);
+  });
+}
+
+function wrapAsyncFunction(f: AsyncExtensionFunction, name: string): AsyncExtensionApplication {
+  return (async args => {
+    let res;
+    try {
+      res = await f(args.map(arg => arg.toRDF()));
+    }catch (e) {
+      throw new ExtensionFunctionError(name, e);
+    }
+    return transformRDFTermUnsafe(res);
+  });
+}
+
 // TODO: Support passing functions to override default behaviour;
-export function transformNamed(expr: Alg.NamedExpression): E.NamedExpression {
+export function transformNamed(expr: Alg.NamedExpression, config: TransformAlgebraConfig)
+    : E.NamedExpression | E.AsyncExtensionExpression | E.SyncExtensionExpression{
   const funcName = expr.name.value;
+  const args = expr.args.map((a) => transformAlgebra(a, config));
+  let simpleAppl: SimpleApplication;
+  let asyncAppl: AsyncExtensionApplication;
   if (!C.NamedOperators.contains(funcName as C.NamedOperator)) {
-    throw new Err.UnknownNamedOperator(expr.name.value);
+    if (!simpleAppl && !asyncAppl && config.syncExtensionFunctionCB) {
+      const func = config.syncExtensionFunctionCB(expr.name);
+      if (func) {
+        simpleAppl = wrapSyncFunction(func, expr.name.value);
+        return new E.SyncExtension(expr.name, args, simpleAppl);
+      }
+    }
+    if (!simpleAppl && !asyncAppl && config.asyncExtensionFunctionCB) {
```

Comment:
We probably want to error early on if we receive both `syncExtensionFunctionCB` and `asyncExtensionFunctionCB`.